code/modules/mob/living/basic/guardian/guardian.dm

```diff
@@ -0,0 +1,332 @@
+	remove_all_languages(LANGUAGE_MASTER)
+	copy_languages(to_who, LANGUAGE_MASTER) // make sure holoparasites speak same language as master
+	RegisterSignal(to_who, COMSIG_QDELETING, PROC_REF(on_summoner_deletion))
+	RegisterSignal(to_who, COMSIG_LIVING_ON_WABBAJACKED, PROC_REF(on_summoner_wabbajacked))
+	RegisterSignal(to_who, COMSIG_LIVING_SHAPESHIFTED, PROC_REF(on_summoner_shapeshifted))
+	RegisterSignal(to_who, COMSIG_LIVING_UNSHAPESHIFTED, PROC_REF(on_summoner_unshapeshifted))
+	recall(forced = TRUE)
+	leash_to(src, summoner)
+	if (to_who.stat == DEAD)
+		on_summoner_death(src, to_who)
+	summoner.updatehealth()
+
+/// Remove all references to our summoner
+/mob/living/basic/guardian/proc/cut_summoner(different_person = FALSE)
+	if (QDELETED(summoner) || isnull(summoner))
+		return
+	if (is_deployed())
+		recall_effects()
+	var/summoner_turf = get_turf(src)
+	if (!isnull(summoner_turf))
+		forceMove(summoner_turf)
+	unleash()
+	UnregisterSignal(summoner, list(COMSIG_QDELETING, COMSIG_LIVING_ON_WABBAJACKED, COMSIG_LIVING_SHAPESHIFTED, COMSIG_LIVING_UNSHAPESHIFTED))
+	if (different_person)
+		summoner.faction -= "[REF(src)]"
+		faction -= summoner.faction
+		mind?.remove_all_antag_datums()
+	if (!length(summoner.get_all_linked_holoparasites() - src))
+		for (var/action_type in control_actions)
+			var/datum/action/remove_action = locate(action_type) in summoner.actions
+			if (isnull(remove_action))
+				continue
+			remove_action.Remove(summoner)
+	summoner = null
+
+/// Connects these two mobs by a leash
+/mob/living/basic/guardian/proc/leash_to(atom/movable/leashed, atom/movable/leashed_to)
+	leashed.AddComponent(\
+		/datum/component/leash,\
+		owner = leashed_to,\
+		distance = range,\
+		force_teleport_out_effect = /obj/effect/temp_visual/guardian/phase/out,\
+		force_teleport_in_effect = /obj/effect/temp_visual/guardian/phase,\
+	)
+
+/// Removes the leash from this guardian
+/mob/living/basic/guardian/proc/unleash()
+	qdel(GetComponent(/datum/component/leash))
+
+/// Called when our owner dies. We fucked up, so now neither of us get to exist.
+/mob/living/basic/guardian/proc/on_summoner_death(mob/living/source, mob/living/former_owner)
+	cut_summoner()
+	if (!isnull(former_owner.loc))
+		forceMove(former_owner.loc)
+	to_chat(src, span_danger("Your summoner has died!"))
+	visible_message(span_bolddanger("\The [src] dies along with its user!"))
+	former_owner.visible_message(span_bolddanger("[former_owner]'s body is completely consumed by the strain of sustaining [src]!"))
+	former_owner.dust(drop_items = TRUE)
+
+/// Called when our health changes, inform our owner of why they are getting hurt (if they are)
+/mob/living/basic/guardian/proc/on_harm(mob/living/source, mob/living/summoner, amount)
+	if (QDELETED(src) || QDELETED(summoner) || amount <= 2)
+		return
+	to_chat(summoner, span_bolddanger("[name] is under attack! You take damage!"))
+	summoner.visible_message(span_bolddanger("Blood sprays from [summoner] as [src] takes damage!"))
```

Comment:
couldn't these both be folded into the same `visible_message()` call? bit cheaper and will let us multiline nicer.

Reply:
No, otherwise blind mobs couldn't tell if they're under attack.